Core/include/Acts/TrackFitting/MbfSmoother.hpp

```diff
@@ -0,0 +1,138 @@
+/// smoothing procedure for a filtered, forward trajectory using the stored
+/// linearization.
+class MbfSmoother {
+ public:
+  struct InternalTrackState {
+    using Projector =
+        typename TrackStateTraits<MultiTrajectoryTraits::MeasurementSizeMax,
+                                  false>::Projector;
+    using Jacobian =
+        typename TrackStateTraits<MultiTrajectoryTraits::MeasurementSizeMax,
+                                  false>::Covariance;
+    using Parameters =
+        typename TrackStateTraits<MultiTrajectoryTraits::MeasurementSizeMax,
+                                  false>::Parameters;
+    using Covariance =
+        typename TrackStateTraits<MultiTrajectoryTraits::MeasurementSizeMax,
+                                  false>::Covariance;
+
+    // This is used to build a covariance matrix view in the .cpp file
+    unsigned int calibratedSize{0};
+    const double* calibrated{nullptr};
+    const double* calibratedCovariance{nullptr};
+    Projector projector;
+
+    Jacobian jacobian{nullptr};
+
+    Parameters predicted{nullptr};
+    Covariance predictedCovariance{nullptr};
+    Parameters filtered{nullptr};
+    Covariance filteredCovariance{nullptr};
+    Parameters smoothed{nullptr};
+    Covariance smoothedCovariance{nullptr};
+
+    InternalTrackState() = default;
+
+    template <typename TrackStateProxy>
+    explicit InternalTrackState(TrackStateProxy ts)
+        : jacobian(ts.jacobian()),
+          predicted(ts.predicted()),
+          predictedCovariance(ts.predictedCovariance()),
+          filtered(ts.filtered()),
+          filteredCovariance(ts.filteredCovariance()),
+          smoothed(ts.smoothed()),
+          smoothedCovariance(ts.smoothedCovariance()) {
+      if (ts.typeFlags().test(TrackStateFlag::MeasurementFlag)) {
+        calibratedSize = ts.calibratedSize();
+        // Note that we pass raw pointers here which are used in the correct
+        // shape later
+        calibrated = {ts.effectiveCalibrated().data()};
+        calibratedCovariance = {ts.effectiveCalibratedCovariance().data()};
+        projector = {ts.projector()};
+      }
+    }
+  };
+
+  /// Run the Kalman smoothing for one trajectory.
+  ///
+  /// @param[in] gctx The geometry context to be used
+  /// @param[in,out] trajectory The trajectory to be smoothed
+  /// @param[in] entryIndex The index of state to start the smoothing
+  /// @param[in] logger Where to write logging information to
+  template <typename traj_t>
+  Result<void> operator()(const GeometryContext& gctx, traj_t& trajectory,
+                          std::size_t entryIndex,
+                          const Logger& logger = getDummyLogger()) const {
+    (void)gctx;
+    (void)logger;
+
+    using TrackStateProxy = typename traj_t::TrackStateProxy;
+
+    TrackStateProxy start_ts = trajectory.getTrackState(entryIndex);
+
+    BoundMatrix big_lambda_hat = BoundMatrix::Zero();
+    BoundVector small_lambda_hat = BoundVector::Zero();
```

Comment:
Do you think we could add a comment with a  link to the notation used for these matrices / vectors?